### PR TITLE
Large Fusion

### PR DESCRIPTION
**Pull request by GlodBlock**

### CODE REVIEW

**Comment by Pxx500** [comment]
Doest this add tier 5 fusion reactor?
It there an issue with more information about this?

**Comment by GTNH-Colen** [comment]
So I can't speak to the programming side much but here are my observations so far that should probably be addressed

- Recipe button on precision assembler is too large in NEI handler. (Don't know when this was added but thought I'd mention it in a ticket as well).
- Add uranium reinforced borosilicate glass to bartworks directly and label it correctly as UHV-tier glass.
- Add fusion casing/coil count to tooltip (1698 and 558 respectively).
- Compact fusion coil MK-II is a 1:1 recipe, is this intended? All other compact coils are 1:2. Perhaps make all coil ratios 1:1? The coil count decreases until the MK4 currently which seems a bit weird. However this would change the pricing by a fair margin.

**Comment by Steelux8** [comment]
The mk4 seems to have a weirdly small number of items needed, is that on purpose?

**Review by GTNH-Colen — CHANGES_REQUESTED** [review]
1698 casings, not 1668. Technically 1666 with 32 energy hatches but it's best if we just keep it as 1698 for simplicity I think.

**Comment by GTNH-Colen** [comment]
Alright so last point from me and this will likely be annoying but I believe Alk has finished the MK5 reactor now (just needs to add a couple recipes). Given the MK5 will replace the MK4 ID wise this will fuck up your MK4 mega fusion. For that reason I think you should adapt to the MK5 and add a new reactor now otherwise it will cause trouble. I just loaded up your world with the fusion changes from GTPP and now there is a MK5 reactor recipe in your MK4 compact fusion for this reason.

Annoying but I think it's best to address this now while it's fresh in your mind and stops conflicts next update (when both mods will be up to date).

**Comment by Houstonruss** [comment]
nocontent
hmm yes

**Comment by GlodBlock** [comment]
alk wrote some mysterious thing in mk4/5 so i have to keep up with them. now compact mk4/5 has the same mechanism like the normal one but haven't fully tested like the compact mk1~3

**Comment by GlodBlock** [comment]
also the recipe related with mk4/5 is still can be balanced at any time cuz alk hasn't add recipe for mk4/5

File: src/main/java/goodgenerator/blocks/tileEntity/LargeFusionComputer4.java [hunk 1]
```diff
@@ -70,17 +77,27 @@ public LargeFusionComputer4(String name) {
...
-        return 1280060000L * (Math.min(32, this.mEnergyHatches.size() + this.eEnergyMulti.size())) / 32L;
+        return 2560060000L * (Math.min(32, this.mEnergyHatches.size() + this.eEnergyMulti.size())) / 32L;
     }
 
     @Override
     public Block getCasingBlock() {
-        return ModBlocks.blockCasings3Misc;
+        try {
+            Field f = ModBlocks.class.getDeclaredField("blockCasings6Misc");
+            return (Block) f.get(null);
```
Comment (draknyte1) [inline src/main/java/goodgenerator/blocks/tileEntity/LargeFusionComputer4.java:87]: You should cache your reflection.

**Comment by GTNH-Colen** [comment]
After playing with it a bit more I think perhaps we should change the parallel of each tier? At present I can make a mega MK1 and make 2kL of helium plasma per tick. That's 163m EU/t at 81,920 EU per L, with inf turbines (highest efficiency of 270%) it's 442m EU/t. Given it uses 558 coils and each reactor uses 32 normally (the cost is slightly higher for Ameliorated coils but not by much and the casings are trivial) it seems a bit OP. 32 * 256 = 8192 coils or a ~94% reduction in cost. The computers are also something we should maybe adjust given you are saving a shit ton of computers here and they are a very large cost especially for the MK4/5. Perhaps a bump to 20 computers at least?

I would recommend the following, 20 parallel for LuV (still saving 13% of coil costs) and double each tier.

MK1: 20
MK2: 40
MK3: 80
MK4: 160
MK5: 320

This incentivises users to actually make the higher tier reactor instead of spamming MK1s and reigns in some of the OP features of the lower tiers. Thoughts? 

Also I would say move the MK1 mega fusion to an assline recipe given the MK1 computer itself uses the assline as do all the subsequent computers. (These should really be done in TT because then they can get proper research requirements with computation). Also the MK4/5 megas should probably use the compact coil of their actual tier in the recipe instead of (tier -1) but these last two points are very minor nit picks. 

Bug wise I think the MK5 has an overflow issue (see negative EU)

Screenshot 2022-02-18 at 00 12 27

It won't run for more than about a second before shutting off doing tin plasma with all UEV energy hatches (and only gets 170x parallel). When running on the MK3 mega has 256x parallel with all UV energy hatches so something is definitely wrong here.

**Comment by GTNH-Colen** [comment]
Actually I just realised the compact MK5 has absolutely insane parallel and 100% needs to be reduced, just using helium plasma in this beast at 8192x parallel with 125L/0.8s yields 5.2B EU/t or 14.1B EU/t with inf turbines. I think the core issue here is that it relies on GT++ logic which is just too nuts. I can't test it because of the above mentioned bug but I think at first sight we should nuke the MK5 parallel and just have yours being the parallel alternative otherwise this is a nightmare to balance properly. @draknyte1 thoughts?

To further incentivise your reactor and handle endgame resource spam I think I will (finally) try add an infinity plasma recipe that is MK5 locked.

**Comment by GlodBlock** [comment]
20 parallel for a such huge structure is bit silly. how about 
MK1: 64 parallel, 16 controller
MK2: 128 parallel, 32 controller
MK3: 256 parallet, 64 controller
as to the compact mk4/5, alk said them are far from finished, so let's keep the current data temporarily and implement them later. anyway i have made them uncraftable with the earlier version gt++. also it won't cause any crash or error with the current gt++ so it is fine to merge without the mk5 pr being merged in gt++.

**Comment by Pxx500** [comment]
> 20 parallel for a such huge structure is bit silly. how about 
>MK1: 64 parallel, 16 controller 
>MK2: 128 parallel, 32 controller 
>MK3: 256 parallet, 64 controller

How about making higher tiers a little better?
MK1: 64 parallel, 32 controller 
MK2: 128 parallel, 48 controller 
MK3: 256 parallet, 64 controller

**Comment by GlodBlock** [comment]
the parallel is the following as @GTNH-Colen suggested
image
the compact controller is nerfed to 48x normal controller.
the compact coil's recipe is unchanged

**Pull request merged**